### DISCUSSION TITLE
Model functions with keyword-only arguments

### DISCUSSION THREAD

**Discussion by schtandard** [lmfit/lmfit-py#975] [General]
Keyword-only arguments to model functions do not work.

```python
import numpy as np
import lmfit

def foo(x, *, a=1, b=2, kind='linear'):
    if kind == 'linear':
        return a + b * x
    elif kind == 'exponential':
        return a * np.exp(b * x)

mdl = lmfit.Model(foo)
# 'a' and 'b' are not parameters.
print(mdl.param_names)

# 'a' and 'b' are not varied.
# 'kind' is ignored.
x = np.linspace(-10, 10, 21)
y = -2 * np.exp(x / 2)
fitres = mdl.fit(y, x=x, kind='exponential')
fitres.plot_fit()

# This fails.
mdl = lmfit.Model(foo, param_names=['a', 'b'])
```

Is this by design? The documentation states

> Keyword arguments to the model function that have numerical values will also be marked as being Parameters, and the supplied numerical value will be used as the default value for that Parameter.

which leads me to expect that keyword-only arguments should work.

**Comment by newville** [comment]
@schtandard  Model functions do not support the bare `*` (or '/') parameters.

``` python

def foo(x, a=1, b=2, kind='linear'):
    if kind == 'linear':
        return a + b * x
    elif kind == 'exponential':
        return a * np.exp(b * x)

mdl = lmfit.Model(foo)
```

will work.

**Reply by schtandard** [reply]
Would you accept a feature request for changing this? I have functions with keyword-only arguments from other contexts that I would like to use as model functions and this forces me to manually make a copy of the function changing only the signature.

**Reply by newville** [reply]
Yes. Sorry for not saying that earlier.  I think the "parse arguments of model function" dates from before the bare `*` or `/` and we just never got around to supporting that.   Raising this an Issue would be helpful (with your example as a test case we should be able to pass).  A PR would be great too.
